src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,840 @@
+                if (!inetAddr.isUnresolved()) {
+                    String ipAddress = inetAddr.getAddress().getHostAddress();
+                    String identity = String.format("%s:%d", ipAddress, inetAddr.getPort());
+                    if (connectedNeighbors.containsKey(identity)) {
+                        log.info("skipping connecting to {} as neighbor is already connected", identity);
+                        reconnectPool.remove(neighborURI);
+                        return;
+                    }
+                }
+
+                initNeighborConnection(neighborURI, inetAddr);
+            } catch (IOException e) {
+                log.warn("unable to build socket for neighbor {}: {}", neighborURI, e.getMessage());
+            }
+        });
+    }
+
+    /**
+     * Initializes a new {@link SocketChannel} to the given neighbor. <br/>
+     * The IP address of the neighbor is removed from the blacklist, added to the whitelist and registered as an allowed
+     * neighbor by its identity.
+     * 
+     * @param neighborURI The {@link URI} of the neighbor to connect to
+     * @param addr        The {@link InetSocketAddress} extracted from the {@link URI}
+     * @throws IOException if initializing the {@link SocketChannel} fails
+     */
+    private void initNeighborConnection(URI neighborURI, InetSocketAddress addr) throws IOException {
+        if (addr.isUnresolved()) {
+            log.warn("unable to resolve neighbor {} to IP address, will attempt to reconnect later", neighborURI);
+            return;
+        }
+
+        String ipAddress = addr.getAddress().getHostAddress();
+
+        // we are overriding a blacklist entry as we are explicitly trying to create a connection
+        hostsBlacklist.remove(ipAddress);
+
+        // allow connections from the given remote IP address to us.
+        // this comes into place if our own initialized connection fails
+        // but afterwards the added neighbor builds a connection to us.
+        hostsWhitelist.add(ipAddress);
+
+        // map the ip address to the domain
+        ipToDomainMapping.put(ipAddress, addr.getHostString());
+
+        // make the identity of the newly added neighbor clear, so that it gets rejected during handshaking
+        // finalisation, in case the communicated server socket port is wrong.
+        allowedNeighbors.add(String.format("%s:%d", addr.getAddress().getHostAddress(), addr.getPort()));
+
+        // init new TCP socket channel
+        SocketChannel tcpChannel = SocketChannel.open();
+        tcpChannel.socket().setTcpNoDelay(true);
+        tcpChannel.socket().setSoLinger(true, 0);
+        tcpChannel.configureBlocking(false);
+        tcpChannel.connect(addr);
+        Neighbor neighbor = new NeighborImpl<>(selector, tcpChannel, addr.getAddress().getHostAddress(), addr.getPort(),
+                txPipeline);
+        neighbor.setDomain(addr.getHostName());
+        tcpChannel.register(selector, SelectionKey.OP_CONNECT, neighbor);
```

Comment:
This isn't a duplicate because here we are registering for `OP_CONNECT` interest, while the other one directly registers for `OP_READ` and `OP_WRITE` (since we got the connection).

Reply:
added a private method